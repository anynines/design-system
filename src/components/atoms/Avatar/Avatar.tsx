import React from 'react'
import styled from 'styled-components'

// I N T E R F A C E
export interface AvatarProps {
  avatar?: string
  className?: string
  level?: number
  rounded?: boolean
  size?: string
  userName?: string
}

export const getInitials = (userName: string): string => {
  const names = userName.split(' ')
  let initials = names[0].substring(0, 1).toUpperCase()

  if (names.length > 1) {
    initials += names[names.length - 1].substring(0, 1).toUpperCase()
  }

  return initials
}

// C O M P O N E N T
export const Avatar: React.FC<AvatarProps> = ({
  avatar,
  children,
  className,
  level,
  rounded = false,
  size = 'sm',
  userName = ''
}) => {
  const renderAvatarContent = (): React.ReactNode => {
    if (children) {
      return children
    }

    return (
      <div data-testid='initials' className='initials'>
        {!avatar && getInitials(userName)}
      </div>
    )
  }

  return (
    <StyledAvatar
      className={`
        avatar-wrapper
        ${size} 
        ${rounded && 'rounded'} 
        ${level && 'level'} 
        ${!avatar && 'initials'} 
        ${className}
      `}
      style={{
        backgroundImage: `${avatar && `url(${avatar})`}`
      }}
      avatar={avatar}
    >
      {renderAvatarContent()}

      {level ? <div className='level'>{level}</div> : null}
    </StyledAvatar>
  )
}

// S T Y L E S
const StyledAvatar = styled.div`
  --size: 3rem;
  --initials-font-size: 2.4em;
  --level-position: -0.3em;
  --level-size: 2em;
  
  display: flex;
  justify-content: center;
  align-items: center;
  margin-right: var(--space-md);
  border-radius: var(--radius);
  border: var(--border);
  background-color: var(--color-dark);
  background-position: center center;
  background-size: cover;
  width: calc(var(--size) * 0.75);
  height: calc(var(--size) * 0.75);
  font-size: var(--text-xxxs);

  &:hover {
    box-shadow: var(--shadow);
  }

  &.rounded {
    border-radius: 100%;

    &:not(.level) {
      overflow: hidden;
    }
  }

  &.md {
    width: var(--size);
    height: var(--size);
    font-size: var(--text-xxs);
  }

  &.lg {
    width: calc(var(--size) * 2);
    height: calc(var(--size) * 2);
    font-size: var(--text-md);
  }

  &.level {
    position: relative;

    .level {
      display: flex;
      justify-content: center;
      align-items: center;
      position: absolute;
      top: var(--level-position);
      left: var(--level-position);
      background-color: var(--color-primary);
      width: var(--level-size);
      height: var(--level-size);
      color: var(--color-white-fix);
      font-weight: var(--font-weight-bd);
      border-radius: 100%;
    }
  }

  &.initials {
    .initials {
      color: var(--color-white-fix);
      font-size: var(--initials-font-size);
      font-weight: var(--font-weight-bd);
    }
  }

  & > *:not(.initials):not(.level) {
    width: 100%;
    height: 100%;
  }
`
StyledAvatar.displayName = 'StyledAvatar'
