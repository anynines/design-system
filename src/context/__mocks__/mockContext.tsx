import React, { ReactNode } from 'react'
import { ThemeProvider } from 'styled-components'
import { render } from '@testing-library/react'

import { GlobalStyle } from '../../designSystemInstance/theme/GlobalStyle'

// D E F A U L T  T H E M E
import { defaultTheme } from '../../designSystemInstance/theme/defaultTheme'

export function renderWithProviders(children: ReactNode) {
  return render(
    <ThemeProvider theme={defaultTheme}>
      <GlobalStyle />
      {children}
    </ThemeProvider>
  )
}
