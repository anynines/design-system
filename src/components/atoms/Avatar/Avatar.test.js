import React from 'react'
import 'jest-styled-components'

import { Avatar, getInitials } from './Avatar'
import { renderWithProviders } from '../../../context/__mocks__/mockContext'

describe('Avatar component', () => {
  let size; let rounded; let sampleAvatarUrl; let sampleUserName; let level; let
    component

  beforeEach(() => {
    size = 'md'
    rounded = true
    sampleUserName = 'Test User with some Subnames'
    sampleAvatarUrl = 'https://images.pexels.com/photos/3290068/pexels-photo-3290068.jpeg?auto=compress&cs=tinysrgb&dpr=2&w=500'
    level = 46
  })

  describe('snapshots', () => {
    it('should match snapshot with image', () => {
      component = renderWithProviders(<Avatar size={size} avatar={sampleAvatarUrl} level={level} rounded={rounded} />)
      const { asFragment } = component
      expect(asFragment()).toMatchSnapshot()
    })

    it('should match snapshot with username', () => {
      component = renderWithProviders(<Avatar size={size} userName={sampleUserName} level={level} rounded={rounded} />)
      const { asFragment } = component
      expect(asFragment()).toMatchSnapshot()
    })
  })

  describe('helpers', () => {
    it('should generate correct initials', () => {
      const expectedVal = 'TS'
      const receivedVal = getInitials(sampleUserName)

      expect(receivedVal).toBe(expectedVal)
    })
  })

  describe('texts', () => {
    it('should print correct label', () => {
      component = renderWithProviders(<Avatar size={size} avatar={sampleAvatarUrl} level={level} />)
      const { container } = component
      const lvl = container.querySelector('.level')
      expect(Number(lvl.textContent)).toBe(level)
    })

    it('should print correct initial', () => {
      component = renderWithProviders(<Avatar size={size} userName={sampleUserName} level={level} rounded={rounded} />)
      const { getByTestId } = component
      const init = getByTestId('initials')
      expect(init.textContent).toBe('TS')
    })
  })
})
