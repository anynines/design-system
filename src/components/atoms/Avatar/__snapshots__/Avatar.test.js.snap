// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar component snapshots should match snapshot with image 1`] = `
<DocumentFragment>
  .c0 {
  --size: 3rem;
  --font-size: .8em;
  --initials-font-size: 2.4em;
  --level-position: -0.3em;
  --level-size: 2em;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-right: var(--space-lgr);
  border-radius: var(--radius);
  border: var(--border);
  background-color: var(--color-dark);
  background-position: center center;
  background-size: cover;
  width: calc(var(--size) * 0.75);
  height: calc(var(--size) * 0.75);
  font-size: var(--text-sm);
}

.c0:hover {
  box-shadow: var(--shadow);
}

.c0.rounded {
  border-radius: 100%;
}

.c0.rounded:not(.level) {
  overflow: hidden;
}

.c0.md {
  width: var(--size);
  height: var(--size);
  font-size: var(--text-sm);
}

.c0.lg {
  width: calc(var(--size) * 2);
  height: calc(var(--size) * 2);
  font-size: var(--text-lg);
}

.c0.level {
  position: relative;
}

.c0.level .level {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: absolute;
  top: var(--level-position);
  left: var(--level-position);
  background-color: var(--color-primary);
  width: var(--level-size);
  height: var(--level-size);
  color: var(--color-white-fix);
  font-size: var(--font-size);
  font-weight: var(--font-weight-bd);
  border-radius: 100%;
}

.c0.initials .initials {
  color: var(--color-white-fix);
  font-size: var(--initials-font-size);
  font-weight: var(--font-weight-bd);
}

.c0 > *:not(.initials):not(.level) {
  width: 100%;
  height: 100%;
}

<div
    class="c0 
        md 
        rounded 
        level 
        false 
        undefined
      "
    style="background-image: url(https://images.pexels.com/photos/3290068/pexels-photo-3290068.jpeg?auto=compress&cs=tinysrgb&dpr=2&w=500);"
  >
    <div
      class="initials"
      data-testid="initials"
    />
    <div
      class="level"
    >
      46
    </div>
  </div>
</DocumentFragment>
`;

exports[`Avatar component snapshots should match snapshot with username 1`] = `
<DocumentFragment>
  .c0 {
  --size: 3rem;
  --font-size: .8em;
  --initials-font-size: 2.4em;
  --level-position: -0.3em;
  --level-size: 2em;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-right: var(--space-lgr);
  border-radius: var(--radius);
  border: var(--border);
  background-color: var(--color-dark);
  background-position: center center;
  background-size: cover;
  width: calc(var(--size) * 0.75);
  height: calc(var(--size) * 0.75);
  font-size: var(--text-sm);
}

.c0:hover {
  box-shadow: var(--shadow);
}

.c0.rounded {
  border-radius: 100%;
}

.c0.rounded:not(.level) {
  overflow: hidden;
}

.c0.md {
  width: var(--size);
  height: var(--size);
  font-size: var(--text-sm);
}

.c0.lg {
  width: calc(var(--size) * 2);
  height: calc(var(--size) * 2);
  font-size: var(--text-lg);
}

.c0.level {
  position: relative;
}

.c0.level .level {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: absolute;
  top: var(--level-position);
  left: var(--level-position);
  background-color: var(--color-primary);
  width: var(--level-size);
  height: var(--level-size);
  color: var(--color-white-fix);
  font-size: var(--font-size);
  font-weight: var(--font-weight-bd);
  border-radius: 100%;
}

.c0.initials .initials {
  color: var(--color-white-fix);
  font-size: var(--initials-font-size);
  font-weight: var(--font-weight-bd);
}

.c0 > *:not(.initials):not(.level) {
  width: 100%;
  height: 100%;
}

<div
    class="c0 
        md 
        rounded 
        level 
        initials 
        undefined
      "
  >
    <div
      class="initials"
      data-testid="initials"
    >
      TS
    </div>
    <div
      class="level"
    >
      46
    </div>
  </div>
</DocumentFragment>
`;
